### PR TITLE
add fastapi parity test

### PR DESCRIPTION
**Pull request by aniketmaurya**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

This PR adds test to ensure LitServe is always performant than vanilla FastAPI. 

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in GitHub issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: .github/workflows/ci-parity.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.10"
+
+      - name: Install LitServe
+        run: |
+          pip --version
+          pip install .  -U -q
+          pip install -r _requirements/test.txt -U -q
+          pip install torchvision jsonargparse -q
```
Comment (Borda) [inline .github/workflows/ci-parity.yml:31]:
```suggestion
          pip install . "torchvision" "jsonargparse" -U -q -r _requirements/test.txt -U -q
```
resolves all compatibilities

File: tests/parity_fastapi/benchmark.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import torch
+from PIL import Image
+import numpy as np
+
+device = "cpu" if torch.cuda.is_available() else "cuda"
+device = "mps" if torch.backends.mps.is_available() else device
+
+rand_mat = np.random.rand(2, 224, 224, 3) * 255
+Image.fromarray(rand_mat[0].astype("uint8")).convert("RGB").save("image1.jpg")
+Image.fromarray(rand_mat[1].astype("uint8")).convert("RGB").save("image2.jpg")
```
Comment (Borda) [inline tests/parity_fastapi/benchmark.py:16]: having this in the run function for accidental importing from this module wont create fake images?

File: tests/parity_fastapi/main.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+device = "cuda" if torch.cuda.is_available() else "cpu"
+device = "mps" if torch.backends.mps.is_available() else device
+
+diff_factor = {
+    "cpu": 1,
+    "gpu": 2,
+    "mps": 1,
+}
```
Comment (Borda) [inline tests/parity_fastapi/main.py:50]: move it to the top as a constant or inside main?

File: tests/parity_fastapi/fastapi_server.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        with torch.inference_mode():
+            outputs = self.model(x)
+            _, predictions = torch.max(outputs, 1)
+        return predictions.item()
+
+
+device = "cuda" if torch.cuda.is_available() else "cpu"
+if torch.backends.mps.is_available():
+    device = "mps"
+api = ImageClassifierAPI(device)
```
[the remark this answers is not in the document]
Comment (aniketmaurya) [reply tests/parity_fastapi/fastapi_server.py:45]: we don't import this. this script is run as a script.

[the remark this answers is not in the document]
Comment (Borda) [reply tests/parity_fastapi/fastapi_server.py:45]: then I would rename it as `fastapi-server.py` to prevent importing

**Review by Borda — APPROVED** [review]

File: tests/parity_fastapi/fastapi_server.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (aniketmaurya) [reply tests/parity_fastapi/fastapi_server.py]: yeah great suggestion! updated.

**Review by awaelchli — APPROVED** [review]

**Pull request merged**